source/core/DualTypeConversion.cpp

```diff
@@ -0,0 +1,339 @@
+            {
+                TypeRef enumSubElement = typeRefX->GetSubElement(0);
+                IntIndex fieldOffsetX = enumSubElement->ElementOffset();
+                AQBlock1* pDataXElement = static_cast<AQBlock1*>(pDataX) + fieldOffsetX;
+                success = Apply(enumSubElement, pDataXElement, typeRefY, pDataY);
+                break;
+            }
+            default:
+                break;
+        }
+        return success;
+    };
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::Apply(StringRef stringRefX, StringRef stringRefY)
+    {
+        SubString subString = stringRefX->MakeSubStringAlias();
+        stringRefY->CopyFromSubString(&subString);
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::Apply(Timestamp* timestampX, Timestamp* timestampY)
+    {
+        *timestampY = *timestampX;
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::Apply(std::complex<Single>* complexSingleX, std::complex<Single>* complexSingleY)
+    {
+        *complexSingleY = *complexSingleX;
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::Apply(std::complex<Double>* complexDoubleX, std::complex<Double>* complexDoubleY)
+    {
+        *complexDoubleY = *complexDoubleX;
+        return true;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeConversion::ApplyBooleans(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY)
+    {
+        *reinterpret_cast<Boolean*>(pDataY) = *static_cast<Boolean*>(pDataX);;
+        return true;
+    }
+
+    template<typename T> void ApplyNumeric(T valueX, TypeRef typeRefY, void* pDataY)
+    {
+        EncodingEnum encodingY = typeRefY->BitEncoding();
+        switch (encodingY) {
+            case kEncoding_UInt:
+                switch (typeRefY->TopAQSize()) {
+                case 0:
+                    break;
+                case 1:
+                    *reinterpret_cast<UInt8*>(pDataY) = static_cast<UInt8>(valueX);
+                    break;
+                case 2:
+                    *reinterpret_cast<UInt16*>(pDataY) = static_cast<UInt16>(valueX);
+                    break;
+                case 4:
+                    *reinterpret_cast<UInt32*>(pDataY) = static_cast<UInt32>(valueX);
+                    break;
+                case 8:
+                    *reinterpret_cast<UInt64*>(pDataY) = static_cast<UInt64>(valueX);
+                    break;
+                }
+                break;
+            case kEncoding_Enum:
+                switch (typeRefY->TopAQSize()) {
+                    case 0:
+                        break;
+                    case 1:
+                    {
+                        Int8 integerValueX = static_cast<Int8>(valueX);
```

Comment:
Please create another template function for this code section that is kind of "duplicated" four times.

Reply:
Done.